pkg/statistics/handle/usage/indexusage/collector.go

```diff
@@ -0,0 +1,225 @@
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/pkg/parser/model"
+	"github.com/pingcap/tidb/pkg/statistics/handle/usage/collector"
+)
+
+// GlobalIndexID is the key type for indexUsageMap.
+type GlobalIndexID struct {
+	TableID int64
+	IndexID int64
+}
+
+// Sample is the data structure to store index usage information.
+type Sample struct {
+	LastUsedAt     time.Time
+	QueryTotal     uint64
+	KvReqTotal     uint64
+	RowAccessTotal uint64
+	// 0, 0-1, 1-10, 10-20, 20-50, 50-100, 100
+	PercentageAccess [7]uint64
+}
+
+var bucketBound = [6]float64{0, 0.01, 0.1, 0.2, 0.5, 1.0}
+
+func getIndexUsageAccessBucket(percentage float64) int {
+	if percentage == 0 {
+		return 0
+	}
+
+	bucket := 0
+	for i := 1; i < len(bucketBound); i++ {
+		if percentage >= bucketBound[i-1] && percentage < bucketBound[i] {
+			bucket = i
+			break
+		}
+	}
+	if percentage == 1.0 {
+		bucket = len(bucketBound)
+	}
+
+	return bucket
+}
+
+// NewSample creates a new data point for index usage.
+func NewSample(queryTotal uint64, kvReqTotal uint64, rowAccess uint64, tableTotalRows uint64) Sample {
+	percentageAccess := [len(bucketBound) + 1]uint64{}
+
+	// if the `tableTotalRows` == 0, record the percentage as 1 and use the last bucket.
+	bucket := len(bucketBound)
+	if tableTotalRows > 0 {
+		rowAccessPercentage := float64(rowAccess) / float64(tableTotalRows)
+		bucket = getIndexUsageAccessBucket(rowAccessPercentage)
+	}
+	percentageAccess[bucket] = 1
+
+	return Sample{
+		QueryTotal:       queryTotal,
+		KvReqTotal:       kvReqTotal,
+		RowAccessTotal:   rowAccess,
+		PercentageAccess: percentageAccess,
+		LastUsedAt:       time.Now(),
+	}
+}
+
+type indexUsage map[GlobalIndexID]Sample
+
+func (m indexUsage) updateByKey(id GlobalIndexID, sample Sample) {
+	item := m[id]
+	item.QueryTotal += sample.QueryTotal
+	item.RowAccessTotal += sample.RowAccessTotal
+	item.KvReqTotal += sample.KvReqTotal
+	for i, val := range sample.PercentageAccess {
+		item.PercentageAccess[i] += val
+	}
+
+	if item.LastUsedAt.Before(sample.LastUsedAt) {
+		item.LastUsedAt = sample.LastUsedAt
+	}
+	m[id] = item
+}
+
+func (m indexUsage) update(tableID int64, indexID int64, sample Sample) {
+	id := GlobalIndexID{TableID: tableID, IndexID: indexID}
+	m.updateByKey(id, sample)
+}
+
+func (m indexUsage) merge(destMap indexUsage) {
+	for id := range destMap {
+		item := destMap[id]
+		m.updateByKey(id, item)
+	}
+}
+
+// Collector represents a data structure to record the index usage for the whole node
+type Collector struct {
+	collector collector.GlobalCollector[indexUsage]
+	mapper    indexUsage
+	sync.RWMutex
+
+	closed bool
+}
+
+// NewCollector create an index usage collector
+func NewCollector() *Collector {
+	iuc := &Collector{
+		mapper: make(indexUsage),
+	}
+	iuc.collector = collector.NewGlobalCollector[indexUsage](iuc.merge)
+
+	return iuc
+}
+
+// GetIndexUsage returns the index usage information
+func (c *Collector) GetIndexUsage(tableID int64, indexID int64) *Sample {
```

Comment:
Is it safe to return a pointer here? What if the caller modify it?

Reply:
Maybe it's safer to assume `Sample` is immutable after creation, and never update it in-place like `Sample.QueryTotal += XX`.

Reply:
Nice catch. The safety seems fine in the collector, but I agree it's more safe / performant to use a `Sample` here. It avoids the potential risk in the future, and also avoids a heap allocation in this function. I've modified this function to use `Sample`.